### PR TITLE
[WIP] Add pagination #122

### LINKED ISSUE #122: Implement reusable pagination and number of results for data tables (with Volunteers)
# Summary
As an admin, I want to be able to browse volunteers page by page, and also display the number of results.

## Things to Consider
Following #94 which adds Stimulus Reflex as a new dependency, we want to add pagination for data tables in separate issue. 

So far pagination is not implemented anywhere, so let's add it to Volunteers table first. 
You may look at this example to get inspired https://expo.stimulusreflex.com/demos/tabular?direction=desc&order_by=name&page=1&query= 

# Criteria for Completion
 - [ ] While on the Volunteers index, the user can click on the page numbers to see next page of records
 - [ ] The page should display the total number of results 

Pagination and show results

### PR DESCRIPTION
**Pull request by matisnape**
Resolves #122 

### Description

Implements pagination together with sorting on Volunteers page.

### Type of change

* New feature (non-breaking change which adds functionality)

### Screenshots

screencast 2021-10-30 23-12-46

### CODE REVIEW

**Comment by matisnape** [comment]
Hey @garettarrowood 
This is my WIP on the subject.

Sorting and pagination works, but unfortunately, due to sort and pagination params not pushed to params (they're not present in the url), you can either sort OR browse through pages, you cannot combine both. 

The solution I was basing is this one https://expo.stimulusreflex.com/demos/tabular?direction=desc&order_by=name&page=1&query=

It uses CableReady to persist params in the URL and doesn't use morphs, which causes the pagination bar to show twice (looks like one is appended to the table) and I'm not sure why, it's too implicit for me. I left the code with CableReady commented out for testing (you can just comment out the morphs and uncomment the other part).

I'll try to figure this out but if you have any ideas, or feedback on the current code - especially on making this generic - I'll be happy to hear them :)

File: app/views/admin/volunteers/_volunteers_table.html.erb [hunk 1]
```diff
@@ -28,7 +26,7 @@
     </tr>
   </thead>
   <tbody>
-    <% volunteers.each do |volunteer| %>
+    <% @volunteers.each do |volunteer| %>
```
Comment (garettarrowood) [inline app/views/admin/volunteers/_volunteers_table.html.erb:29]: Shouldn't need to be an instance var if its passed as a local to the partial, right?

**Review by garettarrowood — COMMENTED** [review]
A few days ago someone contributed changing the arrow styles. It doesn't look like its causing a merge conflict for you, but you may need to pull upstream and see how your changes layer with the merged sort arrow improvements.

**Comment by matisnape** [comment]
@garettarrowood I definitely will rebase master at the end :)

File: .tool-versions [hunk 2]
```diff
@@ -1,3 +1,4 @@
 ruby 3.0.2
 yarn 1.22.15
 nodejs 14.17.6
+redis 6.2.6
```
Comment (garettarrowood) [inline .tool-versions:4]: 👍 😄

**Comment by garettarrowood** [comment]
@matisnape - To get this project deployed, I had to make some updates to the Gemfile and already add redis. Could you rebase when you get a chance?

Anything I can do to help move this work along? Just checking in.

**Comment by garettarrowood** [comment]
Wanted to check in again @matisnape . Do you intend to come back to this? If not, we will soon have to close this PR and reopen the issue. (But definitely use what you've started here as inspiration/reference)

**Comment by matisnape** [comment]
Hi @garettarrowood 
Thanks for checking in and waiting for me! Unfortunately, I won't be able to take a look at this any time soon, so feel free to close it :) 
I will definitely get back to this project next year, so far it's been fun to learn something new and contribute  :)

**Pull request closed without merging**